ext/spl/spl_directory.c

```diff
@@ -1914,6 +1914,9 @@ static zend_result spl_filesystem_file_read_ex(spl_filesystem_object *intern, bo
 	}
 
 	if (!buf) {
+		if (!silent) {
+			zend_throw_exception_ex(spl_ce_RuntimeException, 0, "Cannot read from file %s", ZSTR_VAL(intern->file_name));
+		}
 		return FAILURE;
```

Comment:
That fix looks incorrect, the whole point of this branch is that when EOF is hit at the end of a loop it doesn't return an empty string which doesn't exist.

Can you point out which RETURN_THROWS() assertion is triggered?

Reply:
https://github.com/php/php-src/blob/eb5b12c2255238f99adf1be06c46b7a1276311e4/ext/spl/spl_directory.c#L2213-L2215

The test which I provided in this PR can reproduce this assertion failure...
It looks like we have no test coverage for this case.
For real file, we will get NULL from php_stream_get_line(), and php_stream_eof() returns true only in the next call.